### PR TITLE
fix(network): don't retry the peer on an empty response

### PR DESCRIPTION
**Pull request by joshieDo**
`StateFetcher` will choose the next peer by getting the next idle one with the lowest latency. However, if that very same peer returns us an empty response, we will keep requesting it if it keeps being the one with the lowest latency. This PR tries another peer on an empty response.

Found it out by coming up with a E2E scenario with 3 nodes,  where only one had the chain.

### CODE REVIEW

File: crates/net/network/src/fetch/mod.rs [hunk 1]
```diff
@@ -122,9 +122,19 @@ impl StateFetcher {
...
     fn next_peer(&mut self) -> Option<PeerId> {
+        self.next_peer_with_predicate(|_, _| true)
+    }
+
+    /// Returns the _next_ idle peer that's ready to accept a request,
+    /// prioritizing those with the lowest timeout/latency, alongside a passed predicate.
+    /// Once a peer has been yielded, it will be moved to the end of the map
+    fn next_peer_with_predicate<P>(&mut self, predicate: P) -> Option<PeerId>
+    where
+        P: Fn(&&PeerId, &&Peer) -> bool,
```
Comment (Rjected) [inline crates/net/network/src/fetch/mod.rs:133]: why double `&&` here?

Comment (onbjerg) [inline crates/net/network/src/fetch/mod.rs:133]:
```suggestion
        P: Fn(&PeerId, &Peer) -> bool,
```

File: crates/net/network/src/fetch/mod.rs [hunk 2]
```diff
@@ -236,24 +246,42 @@ impl StateFetcher {
...
-            if peer.state.on_request_finished() && !is_error && !is_likely_bad_response {
-                return self.followup_request(peer_id)
+            send_follow_up &= peer.state.on_request_finished();
+        }
+
+        if let Some(request) = request {
+            if res.as_ref().is_ok_and(|res| res.is_empty()) {
+                // This peer does not have the information we requested, make sure we try another
+                // peer if possible
+                if let Some(new_peer_id) = self.next_peer_with_predicate(|id, _| id != &&peer_id) {
```
Comment (onbjerg) [inline crates/net/network/src/fetch/mod.rs:267]:
```suggestion
                if let Some(new_peer_id) = self.next_peer_with_predicate(|id, _| *id != peer_id) {
```

File: crates/net/network/src/fetch/mod.rs [hunk 3]
```diff
@@ -268,13 +296,32 @@ impl StateFetcher {
...
-            if peer.state.on_request_finished() {
-                return self.followup_request(peer_id)
+            send_follow_up &= peer.state.on_request_finished();
+        }
+
+        if let Some(request) = self.inflight_bodies_requests.remove(&peer_id) {
+            if res.as_ref().is_ok_and(|res| res.is_empty()) {
+                // This peer does not have the information we requested, make sure we try another
+                // peer if possible
+                if let Some(new_peer_id) = self.next_peer_with_predicate(|id, _| id != &&peer_id) {
```
Comment (onbjerg) [inline crates/net/network/src/fetch/mod.rs:308]:
```suggestion
                if let Some(new_peer_id) = self.next_peer_with_predicate(|id, _| *id != peer_id) {
```

**Review by Rjected — COMMENTED** [review]
just have a q about double `&`, otherwise it makes sense to not try to request from a peer that just returned an empty response

**Review by onbjerg — CHANGES_REQUESTED** [review]

File: crates/net/network/src/fetch/mod.rs [hunk 4]
```diff
@@ -236,24 +246,42 @@ impl StateFetcher {
...
         if let Some(peer) = self.peers.get_mut(&peer_id) {
             // If the peer is still ready to accept new requests, we try to send a followup
             // request immediately.
-            if peer.state.on_request_finished() && !is_error && !is_likely_bad_response {
-                return self.followup_request(peer_id)
+            send_follow_up &= peer.state.on_request_finished();
+        }
+
+        if let Some(request) = request {
+            if res.as_ref().is_ok_and(|res| res.is_empty()) {
```
Comment (mattsse) [inline crates/net/network/src/fetch/mod.rs:264]: this is already covered with `is_likely_bad_response` and can be reused here I believe

File: crates/net/network/src/fetch/mod.rs [hunk 5]
```diff
@@ -236,24 +246,42 @@ impl StateFetcher {
...
+                if let Some(new_peer_id) = self.next_peer_with_predicate(|id, _| id != &&peer_id) {
+                    let request = self.prepare_block_request(
+                        new_peer_id,
+                        DownloadRequest::GetBlockHeaders {
+                            request: request.request,
+                            response: request.response,
+                            priority: Priority::Normal,
+                        },
+                    );
+                    return Some(BlockResponseOutcome::Request(new_peer_id, request))
```
Comment (mattsse) [inline crates/net/network/src/fetch/mod.rs:276]: this exits early and never sends the response, the caller won't get a chance to penalize the peer based on the response.

File: crates/net/network/src/fetch/mod.rs [hunk 6]
```diff
@@ -268,13 +296,32 @@ impl StateFetcher {
...
+                if let Some(new_peer_id) = self.next_peer_with_predicate(|id, _| id != &&peer_id) {
+                    let request = self.prepare_block_request(
+                        new_peer_id,
+                        DownloadRequest::GetBlockBodies {
+                            request: request.request,
+                            response: request.response,
+                            priority: Priority::Normal,
+                        },
+                    );
+                    return Some(BlockResponseOutcome::Request(new_peer_id, request))
```
Comment (mattsse) [inline crates/net/network/src/fetch/mod.rs:317]: same here

File: crates/net/network/src/fetch/mod.rs (hunk 4 above)
Comment (mattsse) [inline crates/net/network/src/fetch/mod.rs:264]:
```suggestion
            if res.as_ref().is_ok_and(|res| is_likely_bad_response) {
```

**Review by mattsse — CHANGES_REQUESTED** [review]
conceptually, this makes sense, but there's a smol issue with the order

**Pull request closed without merging**